File: leetcode/Unique Paths.py
class Solution:
    # @return an integer
    def uniquePaths(self, m, n):
        self.table = [[]]*m
        for i in range(m):
            self.table[i] = [None] * n
            
        def fun(mm, nn, m, n):
            if mm == m and nn == n:
                return 1
            if self.table[mm-1][nn-1]:
                return self.table[mm-1][nn-1]
            a, b = 0, 0
            if mm + 1 <= m:
                #down
                a = fun(mm+1, nn, m, n)
            if nn + 1 <= n:
                #right
                b = fun(mm, nn+1, m, n)
            self.table[mm-1][nn-1] = a+b
            return a + b
        if m and n:
            return fun(1, 1, m, n)
        else:
            return 0

File: lintcode/Unique_Paths.cpp
class Solution {
public:
    /**
     * @param n, m: positive integer (1 <= n ,m <= 100)
     * @return an integer
     */
    int uniquePaths(int m, int n) {
        int dp[m][n];
        dp[0][0] = 1;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (0 == i && 0 == j)
                    continue;
                if (i - 1 < 0)
                    dp[i][j] = dp[i][j - 1];
                else if (j - 1 < 0)
                    dp[i][j] = dp[i - 1][j];
                else
                    dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
    
        return dp[m - 1][n - 1];
    }
};